### PR TITLE
[Feature] Logit focal loss

### PR DESCRIPTION
**Pull request by josephsdavid**
It would be quite nice to have focal loss from logits, for numerical stability. This PR implements that! We have logit versions of crossentropy et al, so this i think has precedence! 

### PR Checklist

- [ ] Tests are added
- [ ] Entry in NEWS.md
- [x] Documentation, if applicable

### CODE REVIEW

File: src/losses/functions.jl [hunk 1]
```diff
@@ -603,14 +603,56 @@ function focal_loss(ŷ, y; dims=1, agg=mean, γ=2, ϵ=epseltype(ŷ))
...
+ 0.665241   0.665241   0.665241   0.665241   0.665241
+
+julia> Flux.logit_focal_loss(ŷ, y) ≈ 1.1277571935622628
+true
+```
+
+See also: [`Losses.focal_loss`](@ref)
+
+"""
+function logit_focal_loss(ŷ, y; γ=2.0f0, agg=mean, dims=1, ϵ=epseltype(ŷ))
```
[the remark this answers is not in the document]
Comment (josephsdavid) [reply src/losses/functions.jl:640]: Oh nice idea, can do!

File: src/losses/functions.jl [hunk 2]
```diff
@@ -603,14 +603,56 @@ function focal_loss(ŷ, y; dims=1, agg=mean, γ=2, ϵ=epseltype(ŷ))
...
+    _check_sizes(ŷ, y)
+    logpt = logsoftmax(ŷ; dims=dims)
+    logpt .+= ϵ
+    loss = agg(sum(@. -y * (1 - exp.(logpt))^γ * logpt; dims=dims))
+    return loss
```
[the remark this answers is not in the document]
Comment (josephsdavid) [reply src/losses/functions.jl:645]: awesome!

File: src/losses/functions.jl [hunk 3]
```diff
@@ -603,14 +603,56 @@ function focal_loss(ŷ, y; dims=1, agg=mean, γ=2, ϵ=epseltype(ŷ))
...
+ 0  0  1  0  0
+
+julia> ŷ = reshape(-7:7, 3, 5) .* 1f0
+3×5 Matrix{Float32}:
+ 0.0900306  0.0900306  0.0900306  0.0900306  0.0900306
+ 0.244728   0.244728   0.244728   0.244728   0.244728
+ 0.665241   0.665241   0.665241   0.665241   0.665241
+
+julia> Flux.logit_focal_loss(ŷ, y) ≈ 1.1277571935622628
+true
```
[the remark this answers is not in the document]
Comment (josephsdavid) [reply src/losses/functions.jl:634]: Yeah i still need to work through these tests, did not realize about the docstring tests until after already putting tests elsewhere :) Can do !

**Comment by CarloLucibello** [comment]
Closing as stale: this PR has not seen any new commits in over two years, and Flux has changed substantially since then (Optimisers.jl-based training, the new recurrent layers, Reactant/Enzyme support, etc.), so it can no longer be merged as-is.

This is housekeeping, not a rejection of the idea. If you (or anyone else) would like to pick it up, please feel free to reopen this PR or open a fresh one against the current `master` — the contribution is still appreciated.
